diff --git a/src/Entity/WebformSubmission.php b/src/Entity/WebformSubmission.php
index 342afda..8d948e8 100644
--- a/src/Entity/WebformSubmission.php
+++ b/src/Entity/WebformSubmission.php
@@ -400,6 +400,41 @@ class WebformSubmission extends ContentEntityBase implements WebformSubmissionIn
   }
 
   /**
+   * Returns data which can be used for logging an event.
+   *
+   * For example, if a webform submission has been modified, the data is
+   * what has changed.
+   *
+   * @return array
+   *   Associative array with "is-new" (boolean) and "changed", itself an
+   *   associative array of field keys with their before and after
+   *   values, if possible.
+   */
+  public function logData() {
+    $return = [
+      'is-new' => $this->isNew(),
+      'changed' => [],
+    ];
+    if ($return['is-new']) {
+      return $return;
+    }
+    $original = $this->getOriginalData();
+    $data = $this->getData();
+    foreach ($data as $key => $value) {
+      if (isset($original[$key]) && $original[$key] == $value) {
+        continue;
+      }
+      else {
+        $return['changed'][$key] = [
+          'from' => empty($original[$key]) ? '' : $original[$key],
+          'to' => $value,
+        ];
+      }
+    }
+    return $return;
+  }
+
+  /**
    * {@inheritdoc}
    */
   public function getToken() {
diff --git a/src/WebformSubmissionStorage.php b/src/WebformSubmissionStorage.php
index b3bc4bd..5a48d45 100644
--- a/src/WebformSubmissionStorage.php
+++ b/src/WebformSubmissionStorage.php
@@ -1167,7 +1167,7 @@ class WebformSubmissionStorage extends SqlContentEntityStorage implements Webfor
       'webform_id' => $webform_submission->getWebform()->id(),
       'sid' => $webform_submission->id() ?: NULL,
       'handler_id' => NULL,
-      'data' => [],
+      'data' => $webform_submission->logData(),
       'timestamp' => time(),
     ];
     $values['data'] = serialize($values['data']);
